hazelcast/include/hazelcast/client/serialization/pimpl/compact.h

```diff
@@ -0,0 +1,426 @@
+
+    bool is_field_exists(const std::string& fieldName,
+                         enum pimpl::field_kind kind);
+    const pimpl::field_descriptor& get_field_descriptor(
+      const std::string& field_name) const;
+    const pimpl::field_descriptor& get_field_descriptor(
+      const std::string& field_name,
+      enum pimpl::field_kind field_kind) const;
+    exception::hazelcast_serialization unknown_field_exception(
+      const std::string& field_name) const;
+    exception::hazelcast_serialization unexpected_field_kind(
+      enum pimpl::field_kind field_kind,
+      const std::string& field_name) const;
+    template<typename T>
+    boost::optional<T> get_variable_size(
+      const pimpl::field_descriptor& field_descriptor);
+    template<typename T>
+    boost::optional<T> get_variable_size(const std::string& field_name,
+                                         enum pimpl::field_kind field_kind);
+    template<typename T>
+    T get_variable_as_non_null(const pimpl::field_descriptor& field_descriptor,
+                               const std::string& field_name,
+                               const std::string& method_suffix);
+    static exception::hazelcast_serialization unexpected_null_value(
+      const std::string& field_name,
+      const std::string& method_suffix);
+    size_t read_fixed_size_position(
+      const pimpl::field_descriptor& field_descriptor) const;
+    size_t read_var_size_position(
+      const pimpl::field_descriptor& field_descriptor) const;
+    pimpl::compact_stream_serializer& compact_stream_serializer;
+    serialization::object_data_input& object_data_input;
+    const pimpl::schema& schema;
+    uint32_t data_start_position;
+    uint32_t variable_offsets_position;
+    std::function<int(serialization::object_data_input&, uint32_t, uint32_t)>
+      get_offset;
+};
+
+/**
+ *  Provides means of writing compact serialized fields to the binary data.
+ */
+class HAZELCAST_API compact_writer
+{
+public:
+    /**
+     * Writes a 32-bit two's complement signed integer.
+     *
+     * @param fieldName name of the field.
+     * @param value     to be written.
+     */
+    void write_int32(const std::string& field_name, int32_t value);
+
+    /**
+     * Writes an UTF-8 encoded string.
+     *
+     * @param fieldName name of the field.
+     * @param value     to be written.
+     */
+    void write_string(const std::string& field_name,
+                      const boost::optional<std::string>& value);
+
+private:
+    friend compact_writer pimpl::create_compact_writer(
+      pimpl::default_compact_writer* default_compact_writer);
+    friend compact_writer pimpl::create_compact_writer(
+      pimpl::schema_writer* schema_writer);
+    explicit compact_writer(
+      pimpl::default_compact_writer* default_compact_writer);
+    explicit compact_writer(pimpl::schema_writer* schema_writer);
+    pimpl::default_compact_writer* default_compact_writer;
+    pimpl::schema_writer* schema_writer;
```

Comment:
I understand we get rid of two write_xx methods by doing this. SchemaWriter won't have to implement all compact_writer methods. However, at concept level having default_compact_writer and schema_writer as members of compact_writer seems weird to me.